index.coffee

```diff
@@ -1,60 +1,172 @@
 	# based on the first request, in order to emit them only once.
-	resume: ->
-		@requestOpts.headers ?= {}
-		@requestOpts.headers.range = "bytes=#{@bytesRead}-"
-		@requestModule(@requestOpts)
-		.on 'request', (reqObj) =>
-			if @retries is 0
-				@stream.emit('request', reqObj)
-		.on 'response', (response) =>
+	_request: ->
+		if @response?
+			# the first response is emitted to listeners, so only specify the
+			# Range header on subsequent requests, to avoid causing a 206 response
+			# status code, which may be unexpected.
+			@requestOpts.headers ?= {}
+			@requestOpts.headers.range = "bytes=#{@bytesRead}-"
+
+		initial = @request is null
+		aborted = false
+		errored = false
+
+		@request = @requestModule(@requestOpts)
+		.once 'request', (reqObj) =>
+			if initial
+				@emit('request', reqObj)
+		.once 'response', (response) =>
+			@_follow(response)
 			if response.statusCode >= 400
 				@error = new Error("Request failed with status code #{response.statusCode}")
-				@stream.emit('error', @error)
-			else if @retries is 0
-				@stream.emit('response', response)
-				@contentLength = parseInt(response.headers['content-length'])
+				@_retry() # will abort the request
 		.on 'error', (err) =>
-			@stream.emit('socketError', err)
-			@retryOrEnd()
-		.on('end', => @retryOrEnd())
-		.pipe(@stream, end: false)
+			errored = true
+			@emit('socketError', err)
+			if aborted
+				@_retry()
+		.once 'abort', ->
+			aborted = true
+		.once 'complete', =>
+			if errored or (@bytesTotal? and @bytesRead < @bytesTotal)
+				@_retry()
+			else
+				@_end() # received complete response
+
+	_follow: (response, requestId) ->
+		if not @response?
+			@bytesTotal ?= parseInt(response.headers['content-length'])
+			if response.headers['accept-ranges']?.toLowerCase() != 'bytes'
+				# set the error but don't interrupt the request in case it succeeds
+				# (the common case). on first retry, we'll pick up the error and abort
+				@error = new Error('Server does not support Byte Serving')
+			@response = cloneResponse(response).on('data', @push.bind(this))
+			@emit('response', @response)
+
+		# wrapping the stream in a v3 Readable stream streamlines the emission
+		# of some events and avoids some weird edge-cases
+		wrapLegacyStream(response).pipe(@response, end: false)
 
 	# Decide whether to resume downloading based on number
 	# of bytes downloaded and maximum retries.
-	retryOrEnd: ->
-		if @error?
-			return # we have emitted unrecoverable error (eg bad status code)
-		else if not @contentLength?
-			@stream.error('error', new Error('Cannot resume without Content-length response header'))
-		else if @bytesRead == @contentLength
-			@stream.end()
-		else if @bytesRead > @contentLength
-			@stream.emit('error', new Error('Resumable received more bytes than expected'))
-		else if @retries >= @maxRetries
-			@stream.emit('error', new Error('Maximum retries exceeded.'))
-		else
-			@retries += 1
-			@resume()
+	_retry: ->
+		try
```

Comment:
Why is there a `try` where all we do is throw errors to catch later? It seems kinda stranger and we could just assign to `@error` instead, eg
```
if @error
else if @response? and not @bytesTotal
	@error =  new Error('Cannot resume without Content-Length response header')
else if @bytesTotal and @bytesRead > @bytesTotal
	@error = new Error('Received more bytes than expected!')
else if @retries >= @maxRetries
	@error = new Error('Maximum retries exceeded')

if @error
	@emit('error', @error)
	@abort()
	return